docs/proposals/resource_management.md

```diff
@@ -0,0 +1,387 @@
+Pods (and their containers) belonging to a QoS class will be grouped under these top level QoS cgroups. And all containers in a pod are nested under the pod cgroups.
+
+The proposed cgroups hierarchy would allow for more efficient resource management and lead to improvements in node reliability & stability.
+This would also allow for significant latency optimizations in terms of pod eviction on nodes with the use of pod level resource usage metrics.
+This document provides a basic outline of how we plan to implement and rollout this feature.
+
+
+## Non Goals
+
+- Pod level disk accounting will not be tackled in this proposal.<br>
+- Pod level resource specification in the Kubernetes API will not be tackled in this proposal.
+
+## Motivations
+
+Kubernetes currently supports container level isolation only and lets users specify resource requests/limits on the containers [Compute Resources](../../docs/design/resources.md). Kubelet creates cgroups sandbox (via docker) for all these containers.
+Each pod can belong to a separate Quality of Service class. There are a few shortcomings to the current model.
+ - Existing Quality of Service support does not apply to pods as a whole - On-going work to support pod level eviction using QoS requires all containers in a pod to belong to the same class. By having pod level cgroups, it is easy to track pod level usage and make eviction decisions.
+ - Infrastructure overhead per pod is currently charged to the node. The overhead of running of setting up and managing the pod sandbox is currently accounted to the node. If the pod sandbox is a bit expensive, like in the case of hyper, having pod level accounting becomes critical.
+ - For the docker runtime we have a containerd-shim which is a small library that sits in front of a runtime implementation that allows it to be repartented to init, handle reattach from the caller etc. With pod level cgroups containerd-shim can be charged to the pod instead of the machine.
+ - If a container exits, all its anonymous pages (tmpfs) gets accounted to the machine (root). With pod level cgroups, that usage can be attributed to the pod.
+ - Let containers share resources - with pod level limits, containers can choose to burst on demand and still be capped at their pod limits.
+
+## Design
+
+High level requirements for the design are as follows:
+ - Do not break existing users. Ideally, there should be no changes to the Kubernetes API semantics.
+ - Support multiple cgroup managers - systemd, cgroupfs, etc.
+
+How we intend to achieve these high level goals is covered in greater detail in the Implementation Plan.
+
+### Proposed cgroup hierarchy:
+
+For the initial implementation we will only support limits for cpu and memory resources.<br>
+
+We propose the following hierarchy:
+
+```
+$ROOT
+  |
+  +- Pod1
+  |   |
+  |   +- Container1
+  |   +- Container2
+  |   ...
+  +- Pod2
+  |   +- Container3
+  |   ...
+  +- ...
+  |
+  +- Bu
```

Comment:
this was shorthand for Bustable and BestEffort, agree it should be defined earlier in the spec.